### PR TITLE
[WIP] Feat: Adaptive Service

### LINKED ISSUE #1834: GSoC 2022: Explore a new generation of flexible loadbalance algorithms
Flexible loadbalance is an intelligent cluster scheduling algorithm, which can dynamically allocate traffic according to real-time machine load and hardware configuration, so as to maintain the best performance of the whole system. We will set up a flexible service performance test environment to simulate the real cluster status, including instance capacity expansion, timeout, offline due to failure, etc.

So if you are interested in this project and the objective described in the proposal, please leave comments below so we can further exchange opinions.

Please don't forget to apply for GSoC from its official website when the application slot opens.

### PR DESCRIPTION
**Pull request by CoolIceV**
**What this PR does**: 
Provides a flexible loadbalance in unstable environments.

**Which issue(s) this PR fixes**: 
Fixes #1834

**You should pay attention to items below to ensure your pr passes our ci test**
We do not merge pr with ci tests failed

- [ ] All ut passed (run 'go test ./...' in project root)
- [ ] After go-fmt ed , run 'go fmt project' using goland.
- [ ] Golangci-lint passed, run 'sudo golangci-lint run' in project root.
- [ ] After import formatted, (using [imports-formatter](https://github.com/dubbogo/tools#5-how-to-get-imports-formatter) to run 'imports-formatter .' in project root, to format your import blocks, mentioned in [CONTRIBUTING.md](https://github.com/apache/dubbo-go/blob/master/CONTRIBUTING.md) above) 
- [ ] Your new-created file needs to have [apache license](https://raw.githubusercontent.com/dubbogo/resources/master/tools/license/license.txt) at the top, like other existed file does.
- [ ] All integration test passed. You can run integration test locally (with docker env). Clone our [dubbo-go-samples](https://github.com/apache/dubbo-go-samples) project and replace the go.mod to your dubbo-go, and run 'sudo sh start_integration_test.sh' at root of samples project root. (M1 Slice is not Support)

### CODE REVIEW

**Comment by CoolIceV** [comment]
Now only the `dynamic load balancer` on the consumer side is implemented, which takes into account the recent request success rate and request latency, the effect is as follows.

Flexible services are turned on at the marked line position. After that the QPS of different configured services is no longer the same, and the CPU load and latency of lower configured services are significantly reduced.

qps
cpu
rtt99

The next step is to implement an `adaptive concurrent limiter` on the provider side.

**Comment by justxuewei** [comment]
I am thinking you could submit several PRs, instead of an all-in-one PR. For example: one PR for p2c, one PR for adaptive service.

**Comment by CoolIceV** [comment]
Simple test results for adaptive concurrency limiter.
image

So far, a simple adaptive concurrency limiter has been implemented that automatically limits the concurrency when the CPU load is too high, keeping the number of requests being processed stable.

However, the following issues still need to be fixed.
1. the latency still rises when the QPS is too high and the service still may get crashed.
2. the adaptive policy is too simple to adapt to more complex situations and it needs to be improved with reference to `brpc`

**Comment by CoolIceV** [comment]
> Nit: I am thinking you could submit several PRs, instead of an all-in-one PR. For example: one PR for p2c, one PR for adaptive service.

The commits are a bit confusing right now, I'll sort it out later and break it down into multiple PRs.

File: cluster/cluster/adaptivesvc/cluster_invoker.go [hunk 1]
```diff
@@ -65,45 +64,20 @@ func (ivk *adaptiveServiceClusterInvoker) Invoke(ctx context.Context, invocation
...
+	_ = metrics.SlidingWindowCounterMetrics.SetMethodMetrics(invoker.GetURL(), invocation.MethodName(), metrics.Requests, 1)
 	// invoke
 	invocation.SetAttachment(constant.AdaptiveServiceEnabledKey, constant.AdaptiveServiceIsEnabled)
+	startTime := time.Now().UnixNano()
 	result := invoker.Invoke(ctx, invocation)
-
+	rtt := time.Now().UnixNano() - startTime
 	// if the adaptive service encounters an error, DO NOT
 	// update the metrics.
 	if clsutils.IsAdaptiveServiceFailed(result.Error()) {
```
Comment (justxuewei) [inline cluster/cluster/adaptivesvc/cluster_invoker.go:75]: If the two test branches have the same logic, I think you should merge them.

File: cluster/cluster/adaptivesvc/cluster_invoker.go [hunk 2]
```diff
@@ -65,45 +64,20 @@ func (ivk *adaptiveServiceClusterInvoker) Invoke(ctx context.Context, invocation
...
-		if str, strOK := remainingIface.(string); strOK {
-			remainingStr = str
-		} else if strArr, strArrOK := remainingIface.([]string); strArrOK && len(strArr) > 0 {
-			remainingStr = strArr[0]
-		}
-	}
-	if remainingStr == "" {
-		logger.Errorf("[adasvc cluster] The %s field type of value %v should be string.",
-			constant.AdaptiveServiceRemainingKey, remainingIface)
+	} else if errors.Is(result.Error(), context.DeadlineExceeded) {
```
Comment (justxuewei) [inline cluster/cluster/adaptivesvc/cluster_invoker.go:77]: For the ddl execeeded error, it is worth wrapping a function, called `IsDeadlineExceeded()`, in the `clsutils` package.

File: cluster/loadbalance/p2c/loadbalance.go [hunk 3]
```diff
@@ -89,49 +85,62 @@ func (l *p2cLoadBalance) Select(invokers []protocol.Invoker, invocation protocol
...
+
+func successRateWeight(url *common.URL, methodName string) float64 {
+	requestsIface, _ := metrics.SlidingWindowCounterMetrics.GetMethodMetrics(url, methodName, metrics.Requests)
+	acceptsIface, _ := metrics.SlidingWindowCounterMetrics.GetMethodMetrics(url, methodName, metrics.Accepts)
+
+	requests := metrics.ToFloat64(requestsIface)
+	accepts := metrics.ToFloat64(acceptsIface)
+	r := (1 + accepts) / (1 + requests)
+
+	if r > 1 {
```
Comment (justxuewei) [inline cluster/loadbalance/p2c/loadbalance.go:140]: What would make r greater than 1? IMO, if it does, something is going wrong. So an error output should be dumpped.

**Pull request closed without merging**